src/dfx/src/lib/nns/install_nns.rs

```diff
@@ -1,32 +1,485 @@
+    let downloaded_filename = {
+        let filename = tmp_dir.path().join("wasm");
+        let mut file = fs::File::create(&filename)?;
+        file.write_all(&buffer)?;
+        filename
+    };
+    fs::rename(downloaded_filename, target)?;
+    Ok(())
+}
+
+/// Downloads wasm file from the main IC repo CI.
+#[context("Failed to download {} from the IC CI.", target_name)]
+pub async fn download_ic_repo_wasm(
+    target_name: &str,
+    src_name: &str,
+    ic_commit: &str,
+    wasm_dir: &Path,
+) -> anyhow::Result<()> {
+    fs::create_dir_all(wasm_dir)?;
+    let final_path = wasm_dir.join(format!("{target_name}.wasm"));
+    println!("{final_path:?}");
+    if final_path.exists() {
+        return Ok(());
+    }
+
+    let url_str =
+        format!("https://download.dfinity.systems/ic/{ic_commit}/canisters/{src_name}.wasm.gz");
+    let url = Url::parse(&url_str)?;
+    download_gz(&url, &final_path).await
+}
+
+/// Downloads all the core NNS wasms, excluding only the front-end wasms II and NNS-dapp.
+pub async fn download_nns_wasms(env: &dyn Environment) -> anyhow::Result<()> {
+    // TODO: Include the canister ID in the path.  .dfx/local/wasms/nns/${COMMIT}/....
+    let ic_commit = "3982db093a87e90cbe0595877a4110e4f37ac740"; // TODO: Where should this commit come from?
+    for name in [
+        "registry-canister",
+        "governance-canister_test",
+        "governance-canister",
+        "ledger-canister_notify-method",
+        "ic-icrc1-ledger",
+        "root-canister",
+        "cycles-minting-canister",
+        "lifeline",
+        "sns-wasm-canister",
+        "genesis-token-canister",
+        "identity-canister",
+        "nns-ui-canister",
+    ] {
+        download_ic_repo_wasm(name, name, ic_commit, &nns_wasm_dir(env)).await?;
+    }
+    for (wasm, url) in [(II_WASM, II_URL), (ND_WASM, ND_URL)] {
+        download(&Url::parse(url)?, &nns_wasm_dir(env).join(wasm))
+            .await
+            .map_err(|e| anyhow!("Failed to download {wasm:?}: {e:?}"))?;
+    }
+    Ok(())
+}
+
+/// Gets the local replica URL.  Note: This is not the same as the provider URL.
+///
+/// The replica URL hosts the canister dashboard and is used for installing NNS wasms.
+///
+/// Note: The port typically changes every time `dfx start --clean` is run.
+///
+/// # Errors
+/// - Returns an error if the replica URL could not be found.  Typically this indicates that the local replica
+///   is not running or is running in a different location.
+///
+/// # Panics
+/// This code is not expected to panic.
+pub fn get_replica_url(env: &dyn Environment) -> anyhow::Result<Url> {
+    let port_path = env
+        .get_network_descriptor()
+        .local_server_descriptor
+        .as_ref()
+        .ok_or_else(|| {
+            anyhow!("Could not determine port of the local server.  Is the dfx server running?")
+        })?
+        .replica_port_path();
+    let port = fs::read_to_string(port_path).map(|string| string.trim().to_string())?;
```

Comment:
See get_running_replica_port() in bootstrap.rs for a version that will return the right port whether the replica is running or the emulator

Reply:
Good spot.  I have pulled that function out into a util file to make it accessible and reused it.  It isn't clear to me why it takes two arguments though, given that the latter can seemingly be obtained from the first.